Update linter to be more accurate and find more issues

mkeeler opened the pull request:
* Refectored to not use global vars
* Got rid of some limits like only looking for testing.T use in calls to require
* Refactored to look for the testing.T type instead of variables named “t”

Mainly this was so that we can detect the class of problems where the outer testing.T is passed to helper functions which internally will use the testing.T

mkeeler committed:
Update linter to be more accurate and find more issues

* Refectored to not use global vars
* Got rid of some limits like only looking for testing.T use in calls to require
* Refactored to look for the testing.T type instead of variables named “t”

Mainly this was so that we can detect the class of problems where the outer testing.T is passed to helper functions which internally will use the testing.T

dhiaayachi commented on lint-consul-retry.go:
```diff
@@ -93,42 +90,49 @@ type visitor struct {
 		}
-		if retryDepth > 0 && tCallsFailer(node.Fun) {
+
+		if v.retryDepth > 0 && tCallsFailer(node.Fun) {
```
I think this is not needed anymore as it checks if we call `t.Fail|Error` but we already check if we generally use any variable of type testing.T, so this should fall under it.

mkeeler replied:
It was still needed due to the differences in how the AST represents a method invocation vs a function call and where the `t` will show up. I have refactored the code so that usesTestingT will find it for both cases.

dhiaayachi commented on lint-consul-retry.go:
```diff
@@ -253,16 +257,46 @@ func usesRequire(fun ast.Expr) bool {
+			continue
+		}
+
+		if pkg.Name == "testing" {
```
What about `require.TestingT`, or any interface type that is satisfied by `testing.T` and where we pass a testing.T type to the func? Would we be able to catch that with this?

mkeeler replied:
Using a concrete `testing.T` is always wrong within a retry. The trouble with the `testing.TB` interface and other concrete types that may be used that implement those interfaces is that we cannot know for certain whether the usage is bad. `require.TestingT` is a good example. `require.Equal` (and other assertion methods) will pass the `TestingT` parameter off to `assert.Equal`. This usage of `require.TestingT` as a parameter to another function clearly correct. 

If we had more complicated static analysis we might be able to detect that a require.TestingT was created prior to the retry invocation and thus shouldn't be used either. While it would provide even more protections it gets into the realm of probably not worth it. The utility of this linter is that its relatively simple yet can catch a pretty broad swathe of errors. Its definitely not perfect but without significant investments of time, its probably the best we can do for now.

dhiaayachi commented on lint-consul-retry.go:
```diff
@@ -93,42 +90,49 @@ type visitor struct {
-	if retryDepth > 0 && v.depth <= retryDepth {
-		retryDepth = 0
+	if n != nil {
+		// When called with a non-nil ast.Node, we are delving 1 node deeper into the tree
+		// and therefore should update our tracked depth accordingly.
+		v.depth++
+	} else {
+		// Once a sub-tree of the AST has finished being walked Visit(nil) will be invoked.
+		// There is no need to decrement the depth because we are passing the visitor by
+		// value rather than a reference. The previous visitor on up the stack with the correct
+		// depth would be used for subsequent recursive Visit calls.
```
Thank you for adding those comments! It would have took me a while to understand this 😅

dhiaayachi reviewed (approved):
Thank you @mkeeler! I left few questions but nothing blocking.

mkeeler committed:
Increase accuracy, remove unused code and record call sites where issues are

mkeeler commented:
Latest commit does a few things:

1. Allows linting regular non-test go files. We have some test-only packages meant to be imported by other tests. Its important for the linter to run on these as well.
2. Adds file/line number details to where infractions occur to make them easier to go back and fix.
3. Added more comments
4. Refactored things a bit more including to unify checking for method invocations on a `*testing.T` or passing the `*testing.T` to an unrelated function.


With these changes, the linter found another batch of infractions in the Consul repo.
